### PR TITLE
Add faster-coco-eval metric

### PR DESCRIPTION
**Pull request by MiXaiLL76**
## Motivation

### Faster COCO Eval

<https://github.com/MiXaiLL76/faster_coco_eval>

This package wraps a facebook C++ implementation of COCO-eval operations found in the [pycocotools](https://github.com/cocodataset/cocoapi/tree/master/PythonAPI/pycocotools) package. This implementation greatly speeds up the evaluation time for coco's AP metrics, especially when dealing with a high number of instances in an image.

## Modification

I added a variable (use_faster_coco_eval) to the CocoMetric class that allows you to select an additional type of validation, namely COCOeval_faster

## Use cases (Optional)

For our use case with a test dataset of 5000 images from the coco val dataset.
Testing was carried out using the mmdetection framework and the eval_metric.py script. The indicators are presented below.

Visualization of testing **colab_example.ipynb** available in directory [examples/comparison](https://github.com/MiXaiLL76/faster_coco_eval/blob/main/examples/comparison/mmdet/colab_example.ipynb)
[colab_example.ipynb in google collab](https://colab.research.google.com/drive/1qj392oIU8fmeyIFHCtxCrLA8PQQAMoIs)
Tested with rtmdet model bbox + segm

### Summary for 5000 imgs

| Type | faster-coco-eval | pycocotools | Profit |
| :--- | ---------------: | ----------: | -----: |
| bbox |            5.812 |       22.72 |  3.909 |
| segm |            7.413 |      24.434 |  3.296 |

## Checklist

1. [x] Pre-commit or other linting tools are used to fix the potential lint issues.

```bash
mixaill76@HOME-SERVER:~/mmdetection$ pre-commit run --all-files
flake8...................................................................Passed
isort....................................................................Passed
yapf.....................................................................Passed
trim trailing whitespace.................................................Passed
check yaml...............................................................Passed
fix end of files.........................................................Passed
fix requirements.txt.....................................................Passed
fix double quoted strings................................................Passed
check for merge conflicts................................................Passed
fix python encoding pragma...............................................Passed
mixed line ending........................................................Passed
codespell................................................................Passed
mdformat.................................................................Passed
docformatter.............................................................Passed
check algorithm readme...................................................Passed
check copyright..........................................................Passed
```

2. [x] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.

```bash
mixaill76@HOME-SERVER:~/mmdetection$ python3 -m coverage run --branch --source mmdet -m pytest tests/test_evaluation/test_metrics/test_coco_metric.py 
========================= test session starts ==========================
platform linux -- Python 3.10.12, pytest-7.4.3, pluggy-1.3.0
rootdir: /home/mixaill76/mmdetection
configfile: pytest.ini
plugins: xdoctest-1.1.4, anyio-3.7.1
collected 17 items                                                                                                                                                                                                                                                             

tests/test_evaluation/test_metrics/test_coco_metric.py ................. [100%]

tests/test_evaluation/test_metrics/test_coco_metric.py::TestCocoMetric::test_evaluate_0
  /usr/lib/python3.10/tempfile.py:999: ResourceWarning: Implicitly cleaning up <TemporaryDirectory '/tmp/tmp606joo9e'>
    _warnings.warn(warn_message, ResourceWarning)

tests/test_evaluation/test_metrics/test_coco_metric.py::TestCocoMetric::test_evaluate_1
  /usr/lib/python3.10/tempfile.py:999: ResourceWarning: Implicitly cleaning up <TemporaryDirectory '/tmp/tmpcnbtqpk6'>
    _warnings.warn(warn_message, ResourceWarning)

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
==================== 17 passed, 2 warnings in 3.47s ====================
```
3. [x] The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

**Comment by MiXaiLL76** [comment]
@BIGWangYuDong Hello! I'm waiting for you Yudong Wang

**Comment by vansprosk8** [comment]
Good library to speed up calculations

**Comment by BIGWangYuDong** [comment]
LGTM, kindly ping @hhaAndroid have a look

**Review by BIGWangYuDong — APPROVED** [review]

**Comment by MiXaiLL76** [comment]
Hello @hhaAndroid, I see that you are active on the site, please take a look at this PR

**Comment by MiXaiLL76** [comment]
@RangiLyu Maybe you’re here and the project hasn’t been abandoned yet?

**Comment by GeorgePearse** [comment]
I think some of the core maintainers maybe have slightly left the project? Or have gone on extended leave? Would love to here a response from someone? (and try to find root forwards if they have)

File: mmdet/evaluation/metrics/coco_metric.py [hunk 1]
```diff
@@ -8,6 +8,8 @@
 
 import numpy as np
 import torch
+from faster_coco_eval import COCO as FasterCOCO
```
Comment (hhaAndroid) [inline mmdet/evaluation/metrics/coco_metric.py]: I think it's best to make this dependency optional, to prevent affecting the previous usage pattern.

File: requirements/runtime.txt [hunk 2]
```diff
@@ -1,3 +1,4 @@
+faster-coco-eval
```
Comment (hhaAndroid) [inline requirements/runtime.txt]: I think it's best to make this dependency optional, to prevent affecting the previous usage pattern.

**Comment by hhaAndroid** [comment]
@MiXaiLL76 Thank you for your support

File: mmdet/evaluation/metrics/coco_metric.py (hunk 1 above)
Comment (MiXaiLL76) [reply mmdet/evaluation/metrics/coco_metric.py]: done

File: requirements/runtime.txt (hunk 2 above)
Comment (MiXaiLL76) [reply requirements/runtime.txt]: done

**Comment by MiXaiLL76** [comment]
> @MiXaiLL76 Thank you for your support

Thanks for a good framework!

**Comment by MiXaiLL76** [comment]
@hhaAndroid Hi! Can you update your status?

File: mmdet/evaluation/metrics/coco_metric.py [hunk 3]
```diff
@@ -96,6 +105,10 @@ def __init__(self,
         self.classwise = classwise
         # whether to use multi processing evaluation, default False
         self.use_mp_eval = use_mp_eval
+        # whether to use Faster Coco Eval, default False
+        self.use_faster_coco_eval = use_faster_coco_eval
+        if FasterCOCO is None:
```
Comment (BIGWangYuDong) [inline mmdet/evaluation/metrics/coco_metric.py]: Seems this may get error when the user not installed FasterCOCO

```suggestion
        if self.use_faster_coco_eva:
            assert FasterCOCO is not None, 'faster-coco-eval is not installed'
```

**Comment by MiXaiLL76** [comment]
@hhaAndroid @ZwwWayne Is anyone else working on the project? Maybe we can take it on as support?

**Comment by MiXaiLL76** [comment]
The project is dead.
